compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp

```diff
@@ -229,6 +229,48 @@ static Value distributeVectors(ImplicitLocOpBuilder &b, Value funcH,
   return funcH;
 }
 
+/// Various handles produced by reduction splitting.
+struct ReductionSplitResult {
+  /// Handle to the leading elementwise operation, may be null if no such
+  /// operation is present.
+  Value leadingEltwiseH;
+  /// Handle to the fill operation feeding the init of a higher-rank
+  /// more-parallel reduction.
+  Value splitFillH;
+  /// Handle to the higher-rank more-parallel reduction.
+  Value splitLinalgH;
+  /// Handle to the final reduction.
+  Value combinerH;
+  /// Handle to the original fill operation, may be null if the operation was
+  /// not re-matched.
+  Value originalFillH;
+};
+
+/// Builds transform IR requesting to bubble up the "expand_shape" operation
+/// produced as parent of reduction splitting if necessary for fusion of the
+/// leading elementwise operation.
+static ReductionSplitResult buildExpansionBubbleUp(
```

Comment:
There is such a thing as a generic builder of expansionBubbleOp that follows the same API as other things in this file (e.g. Bufferize) and that is independent of the ops involved. 

I was going to ask to split this in 2 between the 2-line op-independent part and the op-specific part OTOH, there is a bigger picture item that ApplyPatternsOp should be able to preserve handles at least in certain cases.

Can you add a TODO? 
Added a bulletpoint to address this later.

Reply:
The preservation is orthogonal and the splitting is introduced in another PR.